Source/Core/Core/HW/DVDThread.cpp

```diff
@@ -0,0 +1,188 @@
+// Copyright 2015 Dolphin Emulator Project
+// Licensed under GPLv2+
+// Refer to the license.txt file included.
+
+#include <cinttypes>
+#include <mutex>
+#include <thread>
+#include <vector>
+
+#include "Common/ChunkFile.h"
+#include "Common/CommonTypes.h"
+#include "Common/Event.h"
+#include "Common/Flag.h"
+#include "Common/MsgHandler.h"
+#include "Common/Thread.h"
+#include "Common/Timer.h"
+#include "Common/Logging/Log.h"
+
+#include "Core/Core.h"
+#include "Core/CoreTiming.h"
+#include "Core/HW/DVDInterface.h"
+#include "Core/HW/DVDThread.h"
+#include "Core/HW/Memmap.h"
+#include "Core/HW/SystemTimers.h"
+
+#include "DiscIO/Volume.h"
+
+namespace DVDThread
+{
+
+static void DVDThread();
+static bool ReadFromDisc();
+
+static void FinishRead(u64 userdata, int cyclesLate);
+static int s_finish_read;
+
+static std::thread s_DVD_thread;
+static Common::Event s_DVD_thread_start_working;
+static Common::Event s_DVD_thread_not_working;
```

Comment:
I think some kind of semaphore will match better than two events. eg you can map the number of read commands queued for the worker. So WaitUntilIdle will block until this number hits zero. StartRead will increment it. And the worker thread will wait for a number bigger than 0 and decrement it when the work is done.
In the end, a semaphore needs as many ressources as one event (condition variable + mutex), but it's likely faster to call WaitUntilIdle redundantly.

But bad luck, we don't have a semaphore within Common, so feel free to ignore me.